pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,715 @@
+	controllerutil "k8s.io/kubernetes/pkg/controller/util/node"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultClusterCIDRConfigKey   = "kubernetes.io/clusterCIDRConfig"
+	defaultClusterCIDRConfigValue = "default"
+)
+
+// CIDRs are reserved, then node resource is patched with them.
+// multiCIDRNodeReservedCIDRs holds the reservation info for a node.
+type multiCIDRNodeReservedCIDRs struct {
+	nodeReservedCIDRs
+	clusterCIDR *cidrset.ClusterCIDR
+}
+
+type multiCIDRRangeAllocator struct {
+	client clientset.Interface
+	// nodeLister is able to list/get nodes and is populated by the shared informer passed to controller.
+	nodeLister corelisters.NodeLister
+	// nodesSynced returns true if the node shared informer has been synced at least once.
+	nodesSynced cache.InformerSynced
+	// Channel that is used to pass updating Nodes and their reserved CIDRs to the background.
+	// This increases a throughput of CIDR assignment by not blocking on long operations.
+	nodeCIDRUpdateChannel chan multiCIDRNodeReservedCIDRs
+	recorder              record.EventRecorder
+	// Keep a set of nodes that are currently being processed to avoid races in CIDR allocation.
+	lock              *sync.Mutex
+	nodesInProcessing sets.String
+	// CIDRMap maps ClusterCIDRConfig labels to ClusterCIDR objects.
+	CIDRMap map[string][]*cidrset.ClusterCIDR
+	// AllocatedIPv4CIDRs is a map to keep track of allocated CIDRs.
+	AllocatedIPv4CIDRs map[string][]string
+	// AllocatedIPv6CIDRs is a map to keep track of allocated CIDRs.
+	AllocatedIPv6CIDRs map[string][]string
+}
+
+// NewMultiCIDRRangeAllocator returns a CIDRAllocator to allocate CIDRs for node (one for each ip family).
+// Caller must always pass in a list of existing nodes to the new allocator.
+// NodeList is only nil in testing.
+func NewMultiCIDRRangeAllocator(
+	client clientset.Interface,
+	nodeInformer informers.NodeInformer,
+	allocatorParams CIDRAllocatorParams,
+	nodeList *v1.NodeList) (CIDRAllocator, error) {
+	if client == nil {
+		klog.Fatalf("kubeClient is nil when starting CIDR Allocator")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "cidrAllocator"})
+	eventBroadcaster.StartStructuredLogging(0)
+	klog.V(0).Infof("Sending events to api server.")
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: client.CoreV1().Events("")})
+
+	ra := &multiCIDRRangeAllocator{
+		client:                client,
+		nodeLister:            nodeInformer.Lister(),
+		nodesSynced:           nodeInformer.Informer().HasSynced,
+		nodeCIDRUpdateChannel: make(chan multiCIDRNodeReservedCIDRs, cidrUpdateQueueSize),
+		recorder:              recorder,
+		nodesInProcessing:     sets.NewString(),
+		lock:                  allocatorParams.CIDRLock,
+		CIDRMap:               allocatorParams.CIDRMap,
+		AllocatedIPv4CIDRs:    allocatorParams.AllocatedIPv4CIDRs,
+		AllocatedIPv6CIDRs:    allocatorParams.AllocatedIPv6CIDRs,
```

Comment:
note: shared with the other controller https://github.com/kubernetes/kubernetes/pull/108290/files#r836635895